src/OpenTelemetry/Logs/LogRecord.cs

```diff
@@ -79,23 +80,106 @@ public sealed class LogRecord
 
         public string FormattedMessage { get; }
 
+        /// <summary>
+        /// Gets the raw state attached to the log. Set to <see
+        /// langword="null"/> when <see
+        /// cref="OpenTelemetryLoggerOptions.ParseStateValues"/> is enabled.
+        /// </summary>
         public object State { get; }
 
+        /// <summary>
+        /// Gets the parsed state values attached to the log. Set when <see
+        /// cref="OpenTelemetryLoggerOptions.ParseStateValues"/> is enabled
+        /// otherwise <see langword="null"/>.
+        /// </summary>
         public IReadOnlyList<KeyValuePair<string, object>> StateValues { get; }
 
         public Exception Exception { get; }
 
+        internal IExternalScopeProvider ScopeProvider { get; set; }
+
         /// <summary>
         /// Executes callback for each currently active scope objects in order
         /// of creation. All callbacks are guaranteed to be called inline from
         /// this method.
         /// </summary>
+        /// <remarks>
+        /// Note: Scopes are only available during the lifecycle of the log
+        /// message being written. If you need to capture scopes to be used
+        /// later (for example in batching scenarios), call <see
+        /// cref="BufferLogScopes"/> to safely capture the values (incurs
+        /// allocation).
+        /// </remarks>
         /// <typeparam name="TState">State.</typeparam>
         /// <param name="callback">The callback to be executed for every scope object.</param>
         /// <param name="state">The state object to be passed into the callback.</param>
-        public void ForEachScope<TState>(Action<object, TState> callback, TState state)
+        public unsafe void ForEachScope<TState>(LogRecordScopeCallback<TState> callback, TState state)
+        {
+            if (this.bufferedScopes != null)
+            {
+                ScopeForEachState<TState> forEachState = new ScopeForEachState<TState>(callback, state, 0);
+                IntPtr statePointer = new IntPtr(Unsafe.AsPointer(ref forEachState));
```

Comment:
Why use unsafe?

The .NET API is:

```csharp
void ForEachScope<TState>(Action<object, TState> callback, TState state)
```

We would be in better shape if it was:
```csharp
void ForEachScope<TState>(Action<object, TState> callback, ref TState state)
   where TState : struct
```

As it is, you can't pass a mutable struct, which is needed to track depth. What the unsafe does is build that struct on the stack. Then pass a pointer into the method to the stack. Then build a ref to that pointer. Basically doing the `ref TState state`. This is unsafe because if someone took that pointer and tried to access the stack later, it would be bad. But we're not doing that, everything executes inline here.

Reply:
I'm trying to confirm if `forEachState` will be allocated on stack and not moved by GC since I couldn't find any document which provides the guarantee.

If I use GCHandle, there will be exception saying `Object contains non-primitive or non-blittable data`.

Reply:
For example, the following code will generate compile error:

```csharp
ScopeForEachState<TState>* p = stackalloc ScopeForEachState<TState>[1];
```

```error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('LogRecord.ScopeForEachState<TState>')```

Reply:
@reyang I would be lying if I told you my understanding of this is 100% rock-solid 😄 If you can tap someone from runtime team to check this work that would be very helpful.

My _understanding_ is that because `ScopeForEachState` is a value type it will be on the stack. No GC, no moving. It won't go onto the heap unless we box it for some reason, which we aren't doing. So it is on the stack, then we call another method, passing in that stack pointer. In the new method it will still be at the same point on the stack so the pointer should be good.

The error on...

```ScopeForEachState<TState>* p = stackalloc ScopeForEachState<TState>[1];```

...I think is more because the struct contains a delegate & a generic. The compiler doesn't know how much space it needs to allocate for that.

If you have something more primitive like...

```csharp
public struct FixedLengthStruct
{
   public int Field1;
   public long Field2;
}
```

Then it _should_ work fine:

```FixedLengthStruct* p = stackalloc FixedLengthStruct[1];```

Reply:
Your explanation aligns with _my understanding_ (but I'm not 100% if my understanding is solid) 😄

Reply:
> This is unsafe because if someone took that pointer and tried to access the stack later, it would be bad. But we're not doing that, everything executes inline here.

That is one reason it is unsafe, but unfortunately not the only one. Another reason it is unsafe is because it prevents the GC from accurately tracking what objects are allowed to be collected. For example:

````C#
struct State
{
    object o;
}

void Run() 
{
    State s = new State();
    s.o = new object();
    IntPtr p = Unsafe.AsPointer(ref s);
    
    GC.Collect(); // the object in s.o may get collected because at this point in the method there are no
                  // references to it. It's easy to imagine that the local stack variable s should keep it alive
                  // but the JIT can see that there are no reads from s for the remainder of this method
                  // and if the code can't observe it then it doesn't need to be there. Whether or not it is
                  // depends on precise implementation details of the JIT and GC.

    ref State r = ref Unsafe.AsRef<State>(p);
    Console.WriteLine(r.o.ToString());              // r.o could be null or it could be a dangling pointer
                                                    // calling ToString() on it won't be good.
}
````
The code here is doing this pattern, where the object is 'callback'.


If you want state to give you mutable storage but don't want to allocate each time I'd suggest lazily allocating a singleton thread-local state object and then pass that as your TState.

Reply:
@noahfalk Great insight thanks for the help on this!

One tweak to your snippet:

```csharp
struct State
{
    object o;
}

void Run(object callback) 
{
    State s = new State();
    s.o = callback;
    IntPtr p = Unsafe.AsPointer(ref s);
    
    GC.Collect(); 

    ref State r = ref Unsafe.AsRef<State>(p);
    Console.WriteLine(r.o.ToString());
}
```

Basically, callback is passed into the method, not newed inside it. Does that change anything as far as the GC tracking is concerned?

I did consider the thread-local but figured it would add more overhead than avoiding the allocation would be worth. I'll do some benchmarking with it.

Reply:
Glad to help : )

> Basically, callback is passed into the method, not newed inside it. Does that change anything as far as the GC tracking is concerned?

The JIT is free to shorten the lifetime of a parameter the same way it shortens the lifetime of a local variable unfortunately. If the user calls it like this you could still be in trouble:
````C#
Run(new Callback());
````

Reply:
We discussed this on our weekly call today and decided the depth tracking wasn't a compelling enough feature to justify the complexity so I just removed it. Thanks for killing my puppy @noahfalk 🤣 J/k we can add it back on `LogRecordScope` in the future if we decide it makes sense to do so.